### PR TITLE
Adding support for ttnn.layer_norm in TTIR and TTNN

### PR DESCRIPTION
**Pull request by djstefanovicTT**
### Ticket
closes https://github.com/tenstorrent/tt-mlir/issues/5866

### Problem description
There is a an API for ttnn.layer_norm op in tt-metal https://docs.tenstorrent.com/tt-metal/latest/ttnn/ttnn/api/ttnn.layer_norm.html that is not yet implemented.

### What's changed
Added TTIR and TTNN support for this op. Added compiler tests, opmodel tests, builder tests and conversion to EmitC and EmitPy.

### Checklist
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

**Review by odjuricicTT — APPROVED** [review]

**Review by azecevicTT — APPROVED** [review]

File: tools/builder/ttir/ttir_builder.py [hunk 1]
```diff
@@ -11969,6 +11969,78 @@ def rms_norm(
             unit_attrs=unit_attrs,
         )
 
+    def layer_norm(
```
Comment (tapspatel) [inline tools/builder/ttir/ttir_builder.py:12212]: can you refactor this to fit the new builder style ops (see def add as an example), we are moving away from this representation

File: tools/golden/mapping.py [hunk 2]
```diff
@@ -824,6 +824,45 @@ def rms_norm_golden(
     return rms_norm.to(input.dtype)
 
 
+def layer_norm_golden(
```
Comment (tapspatel) [inline tools/golden/mapping.py]: see other refactored golden functions (ttir_add_golden) for how args should be defined and handled

**Review by tapspatel — CHANGES_REQUESTED** [review]

File: lib/Dialect/TTIR/IR/TTIROps.cpp [hunk 3]
```diff
@@ -4971,6 +4971,64 @@ ::mlir::LogicalResult mlir::tt::ttir::RMSNormOp::verify() {
...
 }
 
+//===----------------------------------------------------------------------===//
+// LayerNormOp
+//===----------------------------------------------------------------------===//
+::mlir::LogicalResult mlir::tt::ttir::LayerNormOp::verify() {
+  RankedTensorType inputType = getInput().getType();
+  RankedTensorType outputType = getResult().getType();
+
+  // Input and output must have the same shape.
```
Comment (anuragsingh-tt) [inline lib/Dialect/TTIR/IR/TTIROps.cpp]: NIT: some of these comments aren't necessary as they merely restate what the code is doing in English.

Comment (djstefanovicTT) [reply lib/Dialect/TTIR/IR/TTIROps.cpp]: Removed redundant comments.

File: tools/builder/ttir/ttir_builder.py (hunk 1 above)
Comment (djstefanovicTT) [reply tools/builder/ttir/ttir_builder.py:12212]: Updated.

File: tools/golden/mapping.py (hunk 2 above)
Comment (djstefanovicTT) [reply tools/golden/mapping.py]: Updated.

File: tools/builder/ttir/ttir_builder.py [hunk 4]
```diff
@@ -11969,6 +11969,208 @@ def rms_norm(
...
+        else:
+            mlir_output_type = self._get_type_from_torch_dtype(output_type)
+
+        input0 = self._get_golden_tensor(in0)
+        weight0 = self._get_golden_tensor(weight) if weight is not None else None
+        bias0 = self._get_golden_tensor(bias) if bias is not None else None
+        op_golden_function = get_golden_function(ttir_op)
+        golden_output = op_golden_function(
+            input0,
+            mlir_output_type,
```
Comment (tapspatel) [inline tools/builder/ttir/ttir_builder.py]: put mlir output type at the end of the list, we will refactor this to communalize code soon

File: tools/golden/mapping.py [hunk 5]
```diff
@@ -824,6 +824,26 @@ def rms_norm_golden(
     return rms_norm.to(input.dtype)
 
 
+def ttir_layer_norm_golden(
+    input: GoldenMapTensor,
+    output_type_mlir: Type,
+    weight: Optional[GoldenMapTensor] = None,
+    bias: Optional[GoldenMapTensor] = None,
+    normalized_shape: List[int] = None,
+    epsilon: float = 1e-5,
```
Comment (tapspatel) [inline tools/golden/mapping.py]: change these to mlir attributes, not python type, reason being we will add passes that will call these golden functions and pass through mlir types

File: tools/builder/ttir/ttir_builder.py (hunk 4 above)
Comment (djstefanovicTT) [reply tools/builder/ttir/ttir_builder.py]: Updated.

File: tools/golden/mapping.py (hunk 5 above)
Comment (djstefanovicTT) [reply tools/golden/mapping.py]: Updated.

File: tools/builder/ttir/ttir_builder.py [hunk 6]
```diff
@@ -12047,6 +12047,208 @@ def rms_norm(
...
+                result.element_type,
+                weight=weight0,
+                bias=bias0,
+                normalized_shape=normalized_shape,
+                epsilon=epsilon,
+            )
+            self._set_golden_tensor(new_op_result, golden_output)
+
+        op_map_dictionary = {}
+        op_map_dictionary[old_op.result] = new_op_result
```
Comment (anuragsingh-tt) [inline tools/builder/ttir/ttir_builder.py]: NIT: these two statements can be one line.

File: tools/builder/ttir/ttir_builder.py [hunk 7]
```diff
@@ -12047,6 +12047,208 @@ def rms_norm(
...
+                        )
+                        layer_norm_builder._set_golden_tensor(
+                            new_op_result, golden_output
+                        )
+                        layer_norm_builder._set_golden_tensor(in0, input0)
+                        ordered_inputs.append(in0)
+                        if weight is not None:
+                            layer_norm_builder._set_golden_tensor(weight, weight0)
+                            ordered_inputs.append(weight)
+                        if bias is not None:
```
Comment (anuragsingh-tt) [inline tools/builder/ttir/ttir_builder.py:12391]: NIT: this can be a for loop instead.

File: tools/builder/ttir/ttir_builder.py [hunk 8]
```diff
@@ -12047,6 +12047,208 @@ def rms_norm(
...
+    ) -> Tuple[Module, TTIRBuilder]:
+        ttir_op = self.get_opview_from_split(TTIRBuilder.layer_norm_split)
+
+        old_ctx = old_op.context
+        old_loc = Location.unknown(old_ctx)
+        with old_ctx, old_loc:
+            layer_norm_module = Module.create()
+            layer_norm_builder = TTIRBuilder(old_ctx, old_loc)
+            op_input_types = [old_op.input.type]
+            if old_op.weight:
```
Comment (anuragsingh-tt) [inline tools/builder/ttir/ttir_builder.py]: NIT: use `is not None`.

Comment (djstefanovicTT) [reply tools/builder/ttir/ttir_builder.py]: Updated.

File: tools/builder/ttir/ttir_builder.py (hunk 6 above)
Comment (djstefanovicTT) [reply tools/builder/ttir/ttir_builder.py]: Updated.

File: tools/builder/ttir/ttir_builder.py (hunk 7 above)
Comment (djstefanovicTT) [reply tools/builder/ttir/ttir_builder.py:12391]: I do not really see why to convert this to for loop. There are multiple variables (weight and weight0 / bias and bias0) and nesting all within a loop would make it less readable and just add a level of indentation.

File: tools/builder/ttir/ttir_builder.py [hunk 9]
```diff
@@ -12047,6 +12047,207 @@ def rms_norm(
...
+        if not self._disable_golden_check:
+            input0 = self._get_golden_tensor(in0)
+            weight0 = self._get_golden_tensor(weight) if weight is not None else None
+            bias0 = self._get_golden_tensor(bias) if bias is not None else None
+            normalized_shape = list(normalized_shape_attr)
+            epsilon = float(epsilon_attr.value)
+            op_golden_function = get_golden_function(ttir_op)
+            golden_output = op_golden_function(
+                input0,
+                result.element_type,
```
Comment (tapspatel) [inline tools/builder/ttir/ttir_builder.py]: this should be the last parameter in the list

File: tools/builder/ttir/ttir_builder.py [hunk 10]
```diff
@@ -12047,6 +12047,207 @@ def rms_norm(
...
+
+        input0 = self._get_golden_tensor(in0)
+        weight0 = self._get_golden_tensor(weight) if weight is not None else None
+        bias0 = self._get_golden_tensor(bias) if bias is not None else None
+        op_golden_function = get_golden_function(ttir_op)
+        golden_output = op_golden_function(
+            input0,
+            weight=weight0,
+            bias=bias0,
+            normalized_shape=normalized_shape,
```
Comment (tapspatel) [inline tools/builder/ttir/ttir_builder.py]: pass the mlir type `normalized_shape_attr`

File: tools/builder/ttir/ttir_builder.py [hunk 11]
```diff
@@ -12047,6 +12047,207 @@ def rms_norm(
...
+            epsilon=epsilon,
```
Comment (tapspatel) [inline tools/builder/ttir/ttir_builder.py]: pass the mlir type `epsilon_attr`

File: tools/builder/ttir/ttir_builder.py [hunk 12]
```diff
@@ -12047,6 +12047,207 @@ def rms_norm(
...
+            bias0 = self._get_golden_tensor(bias) if bias is not None else None
+            normalized_shape = list(normalized_shape_attr)
+            epsilon = float(epsilon_attr.value)
+            op_golden_function = get_golden_function(ttir_op)
+            golden_output = op_golden_function(
+                input0,
+                result.element_type,
+                weight=weight0,
+                bias=bias0,
+                normalized_shape=normalized_shape,
```
Comment (tapspatel) [inline tools/builder/ttir/ttir_builder.py]: pass the mlir type `normalized_shape_attr`

File: tools/builder/ttir/ttir_builder.py [hunk 13]
```diff
@@ -12047,6 +12047,207 @@ def rms_norm(
...
+                epsilon=epsilon,
```
Comment (tapspatel) [inline tools/builder/ttir/ttir_builder.py]: pass the mlir type `epsilon_attr`

File: tools/builder/ttir/ttir_builder.py [hunk 14]
```diff
@@ -12047,6 +12047,207 @@ def rms_norm(
...
+                            self._get_golden_tensor(old_op.weight)
+                            if old_op.weight is not None
+                            else None
+                        )
+                        bias0 = (
+                            self._get_golden_tensor(old_op.bias)
+                            if old_op.bias is not None
+                            else None
+                        )
+                        normalized_shape = list(old_op.normalized_shape)
```
Comment (tapspatel) [inline tools/builder/ttir/ttir_builder.py]: no need to convert to python type, just pass the mlir type directly to golden function

File: tools/builder/ttir/ttir_builder.py [hunk 15]
```diff
@@ -12047,6 +12047,207 @@ def rms_norm(
...
+                        epsilon = float(old_op.epsilon.value)
```
Comment (tapspatel) [inline tools/builder/ttir/ttir_builder.py]: same here

File: tools/builder/ttir/ttir_builder.py [hunk 16]
```diff
@@ -12047,6 +12047,207 @@ def rms_norm(
...
+
+                        op_golden_function = get_golden_function(ttir_op)
+                        golden_output = op_golden_function(
+                            input0,
+                            result.element_type,
```
Comment (tapspatel) [inline tools/builder/ttir/ttir_builder.py]: move this to end of parameter list

File: tools/builder/ttir/ttir_builder.py [hunk 17]
```diff
@@ -12047,6 +12047,207 @@ def rms_norm(
...
+                            weight=weight0,
+                            bias=bias0,
+                            normalized_shape=normalized_shape,
+                            epsilon=epsilon,
```
Comment (tapspatel) [inline tools/builder/ttir/ttir_builder.py]: mlir types

```
normalized_shape=normalized_shape,
                            epsilon=epsilon,
```

File: tools/golden/mapping.py [hunk 18]
```diff
@@ -826,6 +826,28 @@ def rms_norm_golden(
     return rms_norm.to(input.dtype)
 
 
+def ttir_layer_norm_golden(
+    input: GoldenMapTensor,
+    weight: Optional[GoldenMapTensor] = None,
+    bias: Optional[GoldenMapTensor] = None,
+    normalized_shape: ArrayAttr = None,
+    epsilon: FloatAttr = None,
+    output_type_mlir: Type = None,
```
Comment (tapspatel) [inline tools/golden/mapping.py]:
```
 weight: Optional[GoldenMapTensor] = None,
    bias: Optional[GoldenMapTensor] = None,
    normalized_shape: ArrayAttr = None,
    epsilon: FloatAttr = None,
    output_type_mlir: Type = None,
```

it looks like we need these values, don't default to None

File: tools/golden/mapping.py [hunk 19]
```diff
@@ -5163,6 +5186,7 @@ def debug_memory_snapshot_golden(
     # Complex operations
     ttnn.MatmulOp: ttnn_matmul_golden,
     ttnn.LinearOp: ttnn_linear_golden,
+    ttnn.LayerNormOp: ttir_layer_norm_golden,
```
Comment (tapspatel) [inline tools/golden/mapping.py]: can you create a copy of `ttnn_layer_norm_golden`? Reason being we will refactor into standalone golden libraries in the future
